### PR TITLE
update notebooks

### PR DESCRIPTION
**Pull request by gf712**
- Various fixes to make python notebooks work
- Added free functions to go around states of objects with new API
- Added `structured_model` and `factor_type` factories for StructuredModel
- Added CI job and script to run all notebooks (to replace the buildbot job)

### CODE REVIEW

**Review by karlnapf — COMMENTED** [review]
This is grep right?
I think some of them will not run without errors in Py3, or did you run them?

**Comment by gf712** [comment]
> This is grep right?
> I think some of them will not run without errors in Py3, or did you run them?

I fixed the notebook job on buildbot so I am running stuff until all the notebooks are functional. But yes, I just did `grep` and I am just running through the PR until it works...

**Comment by karlnapf** [comment]
great initiative! :)

**Comment by gf712** [comment]
> great initiative! :)

Thanks! I have slowly made through all the buildbot jobs... Hopefully all will be done soon and we can release everything working :)

File: src/shogun/regression/LeastAngleRegression.cpp [hunk 1]
```diff
@@ -429,9 +429,6 @@ SGMatrix<ST> LeastAngleRegression::cholesky_delete(SGMatrix<ST>& R, int32_t i_ki
 	return nR;
 }
 
-template bool LeastAngleRegression::train_machine_templated<floatmax_t>(std::shared_ptr<DenseFeatures<floatmax_t>> data);
-template bool LeastAngleRegression::train_machine_templated<float64_t>(std::shared_ptr<DenseFeatures<float64_t>> data);
-template bool LeastAngleRegression::train_machine_templated<float32_t>(std::shared_ptr<DenseFeatures<float32_t>> data);
```
Comment (gf712) [inline src/shogun/regression/LeastAngleRegression.cpp:434]: pretty sure these templates are instantiated by CRTP

File: doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb [hunk 2]
```diff
@@ -118,18 +114,22 @@
...
+    "net_no_reg.put(\"seed\", 10)\n",
     "net_no_reg.quick_connect()\n",
     "net_no_reg.initialize_neural_network()\n",
     "\n",
     "net_l2 = NeuralNetwork(layers)\n",
+    "net_l2.put(\"seed\", 10)\n",
     "net_l2.quick_connect()\n",
     "net_l2.initialize_neural_network()\n",
     "\n",
     "net_l1 = NeuralNetwork(layers)\n",
```
Comment (karlnapf) [inline doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb]: if you touch bits like this, and the ctors are already available through factories, it would be great if you could replace them, e.g. https://github.com/shogun-toolbox/shogun/blob/develop/examples/meta/src/neural_nets/feedforward_net_classification.sg.in

Of course, only if this works out of the box (e.g. when there are examples to copy paste it from).and you have the nerves ;)

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb]: Ah cool, I thought it wasn't working properly with factories, because the old API is still here. I'll have a look. Now the CI is finally actually running the job.

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb]: actually this is not being run by the interfaces...

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb]: @karlnapf How do you think we should expose `DeepAutoencoder::pre_train(std::shared_ptr<Features>)`?

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb]: and DeepAutoencoder::std::shared_ptr<DenseFeatures<float64_t>> data>)?

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb]: @vigsterkr thoughts? ^

Comment (vigsterkr) [reply doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb]: ain't got no DenseFeatures in SWIG :(

Comment (karlnapf) [reply doc/ipython-notebooks/neuralnets/neuralnets_digits.ipynb]: this is one of the cases where it is a bit of work. We need to provide an abstraction for such operations in the base api. So far, we managed to work around this, via something like
```
nn.put('pre_train_features', f)
nn.run('pre_train') # throws error if above wasnt done by the user
```
This is also done for things like labels (runtime error if not set), kernels, and more specific situations. I am not the biggest fan (especially for labels, which imo should be passed to `::train`, but that is another discussion), but there *will* be methods that are specific to particular classes and wont be in the base api.

Another solution could be a `watch_method` with a single sgobject parameter (see our discussion about adding one with an index), which is then dynamically casted to the desired class.
```
nn.run('pre_train', f)
```

But extending the base API in this matter is also at risk of making things complicated.

The root problem really is the design of the NN class here. Under Shogun's paradigm, it is bad style of relying on such mechanics. So a re-design of the class would be best. Where something like "pre-train" is not necessary, or called inside `::train`

File: doc/ipython-notebooks/classification/MKL.ipynb [hunk 3]
```diff
@@ -694,30 +685,30 @@
...
     "\n",
-    "mkl = sg.MKLMulticlass(1.2, kernel, labels)\n",
+    "mkl = sg.machine(\"MKLMulticlass\", C=1.2, kernel=kernel, \n",
+    "                 labels=labels, mkl_eps=0.001, mkl_norm=1)\n",
     "\n",
-    "mkl.set_epsilon(1e-2)\n",
-    "mkl.set_mkl_epsilon(0.001)\n",
-    "mkl.set_mkl_norm(1)\n",
+    "# FIX ME: need a way to interact with SVM\n",
+    "# mkl.set_epsilon(1e-2)\n",
```
Comment (gf712) [inline doc/ipython-notebooks/classification/MKL.ipynb]: @karlnapf @vigsterkr need to come up with a way to handle these situations. Would we .get() the SVM and manipulate it?

File: src/shogun/classifier/mkl/MKLMulticlass.cpp [hunk 4]
```diff
@@ -34,9 +32,20 @@ MKLMulticlass::MKLMulticlass(float64_t C, std::shared_ptr<Kernel> k, std::shared
...
+{
 	mkl_eps=0.01;
 	max_num_mkl_iters=999;
 	pnorm=1;
+
+	SG_ADD(&mkl_eps, "mkl_eps", "MKL Epsilon");
+	SG_ADD(&max_num_mkl_iters, "max_num_mkl_iters", "MKL max number of iterations");
+	SG_ADD(
+	    &pnorm, "mkl_norm", "MKL norm", ParameterProperties::NONE | ParameterProperties::CONSTRAIN,
+	    SG_CONSTRAINT(greater_than_or_equal(1.0)));
```
Comment (gf712) [inline src/shogun/classifier/mkl/MKLMulticlass.cpp:48]: @karlnapf added this to MKLMulticlass, should be fine no? Nothing was being exposed to parameter framework.

File: src/shogun/machine/FeatureDispatchCRTP.h [hunk 5]
```diff
@@ -35,7 +35,7 @@ namespace shogun
 	protected:
 		virtual bool train_dense(std::shared_ptr<Features> data)
 		{
-			auto this_casted = this->template as<P>();
+			auto* this_casted = static_cast<P*>(this);
```
Comment (gf712) [inline src/shogun/machine/FeatureDispatchCRTP.h:38]: no need for dynamic casting here

File: src/shogun/regression/LinearRidgeRegression.cpp [hunk 6]
```diff
@@ -39,12 +39,13 @@ void LinearRidgeRegression::init()
 
 	SG_ADD(&m_tau, "tau", "Regularization parameter", ParameterProperties::HYPER);
 	SG_ADD(
-	    &m_use_bias, "use_bias", "Whether or not to fit an offset term");
+	    &m_use_bias, "use_bias", "Whether or not to fit an offset term", 
+	    ParameterProperties::SETTING);
```
Comment (gf712) [inline src/shogun/regression/LinearRidgeRegression.cpp:43]: @karlnapf there is this issue now with crossvalidation that if we don't set a parameter to either hyperparameter or setting it won't be copied/cloned... Need to continue going through all the parameters in all algos and set something..

File: src/shogun/distributions/KernelDensity.cpp [hunk 7]
```diff
@@ -91,12 +91,15 @@ bool KernelDensity::train(std::shared_ptr<Features> data)
 	return true;
 }
 
-SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<DenseFeatures<float64_t>>& test, int32_t leaf_size)
+SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<Features>& test, int32_t leaf_size)
 {
 	require(test,"data not supplied");
 
+	auto dense_feat = std::dynamic_pointer_cast<DenseFeatures<float64_t>>(test);
+	require(dense_feat, "Expected Features to be of type DenseFeatures.");
```
Comment (gf712) [inline src/shogun/distributions/KernelDensity.cpp]: @karlnapf is it fine to have this userfacing?

File: doc/ipython-notebooks/classification/MKL.ipynb (hunk 3 above)
Comment (karlnapf) [reply doc/ipython-notebooks/classification/MKL.ipynb]: it is commented out here, so it didnt work before I guess? :)
Is this something that would need to be applied recursively? I.e. all svms need to have set the epsilon. Or is it something of the MKL class?

File: src/shogun/classifier/mkl/MKLMulticlass.cpp (hunk 4 above)
Comment (karlnapf) [reply src/shogun/classifier/mkl/MKLMulticlass.cpp:48]: yep, otherwise it is also not serialized and loading will reset it to the default setting

File: src/shogun/regression/LinearRidgeRegression.cpp (hunk 6 above)
Comment (karlnapf) [reply src/shogun/regression/LinearRidgeRegression.cpp:43]: yes, I think this whitelisting idea is maybe not the best. What about blacklisting parameters instead in xval? Like readonly ones?

File: src/shogun/distributions/KernelDensity.cpp (hunk 7 above)
Comment (karlnapf) [reply src/shogun/distributions/KernelDensity.cpp]: can't this be moved to the crtp dispatcher? There is an error msg in there. But I guess it is some work. So then it is fine like this as well. Although people will not understand it. I also don't like using classnames in user-facing msgs (as they tend to change). What about say "dense real features"

File: doc/ipython-notebooks/classification/MKL.ipynb (hunk 3 above)
Comment (gf712) [reply doc/ipython-notebooks/classification/MKL.ipynb]: isn't there only a single SVM? In any case, what happens is that `MulticlassSVM::set_epsilon` is called, which casts the underlying `m_machine` to SVM and calls `set_epsilon` on that

File: src/shogun/distributions/KernelDensity.cpp (hunk 7 above)
Comment (gf712) [reply src/shogun/distributions/KernelDensity.cpp]: Hmm it would be some work, because this is not the train method. We would need a dispatcher for get_log_density?

File: doc/ipython-notebooks/classification/MKL.ipynb (hunk 3 above)
Comment (karlnapf) [reply doc/ipython-notebooks/classification/MKL.ipynb]: I see. Maybe
`mkl.get('svm').put('epsilon')`

Since we are using oop design, and the user provides an object, why not do that?

Or we somehow take over the SVM parameters with the MKL class so that we can use "put"?

File: src/shogun/distributions/KernelDensity.cpp (hunk 7 above)
Comment (karlnapf) [reply src/shogun/distributions/KernelDensity.cpp]: ah I see, maybe later right? :)

File: doc/ipython-notebooks/classification/MKL.ipynb (hunk 3 above)
Comment (gf712) [reply doc/ipython-notebooks/classification/MKL.ipynb]: Yup, that's what I did, just get.("machine").put(), it keeps the interface cleaner imo

File: src/shogun/regression/LinearRidgeRegression.cpp (hunk 6 above)
Comment (gf712) [reply src/shogun/regression/LinearRidgeRegression.cpp:43]: Yup, that is probably better... Just need to label all the labels/features as readonly

File: src/shogun/distributions/KernelDensity.cpp (hunk 7 above)
Comment (gf712) [reply src/shogun/distributions/KernelDensity.cpp]: indeed.. Otherwise there will be too much in this PR

File: src/shogun/regression/LinearRidgeRegression.cpp (hunk 6 above)
Comment (karlnapf) [reply src/shogun/regression/LinearRidgeRegression.cpp:43]: ah no! Thats why we did it this way. Features should not be readonly.
We need to whitelist as otherwise we have no way to distinguish between model parameters and things like feautres. I remember now.
I put a PR in to test exactly such things (explicit xval by hand against the class we have), that they have the same results. If we combine this with random parameter content, this will catch missing parameter registrations

File: doc/ipython-notebooks/gaussian_process/variational_classifier.ipynb [hunk 8]
```diff
@@ -1279,7 +1278,7 @@
     "    name=inf.get_name()\n",
     "    prbs=gpc.get_probabilities(features_boundary)\n",
     "    plot_helper(plots[0],x1_boundary,x2_boundary,prbs,n_boundary,name,x1s,x2s,\"without optimizing %d inducing points\"%n_inducing) \n",
-    "    x_inducing=inf.get_inducing_features().get_computed_dot_feature_matrix()\n",
+    "    x_inducing=inf.get_inducing_features().get(\"feature_matrix\")  \n",
```
Comment (gf712) [inline doc/ipython-notebooks/gaussian_process/variational_classifier.ipynb:1280]: @vigsterkr how do you call `get_computed_dot_feature_matrix` with a Features object?

Comment (gf712) [reply doc/ipython-notebooks/gaussian_process/variational_classifier.ipynb:1280]: pretty sure what I wrote here isn't right

Comment (karlnapf) [reply doc/ipython-notebooks/gaussian_process/variational_classifier.ipynb:1280]: have a look at the plot, that makes it clear
I actually think this might work

File: doc/ipython-notebooks/neuralnets/autoencoders.ipynb [hunk 9]
```diff
@@ -391,7 +399,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "conv_nn = ae.convert_to_neural_network(NeuralSoftmaxLayer(10))\n",
+    "conv_nn = sg.convert_to_neural_network(ae, sg.layer(\"NeuralSoftmaxLayer\", num_neurons=10), 0.01)\n",
```
Comment (gf712) [inline doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: @karlnapf @vigsterkr would this be a reasonable tradeoff? Imo much more readable than putting an output layer and then calling get() to get a net Machine out...

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: also, each free function adds about 100 LoC in SWIG

Comment (karlnapf) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: I say do it.

But one question: Could we maybe overload the factory for this?

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: Hmmm not sure I understand. You mean the cast functions?

Comment (karlnapf) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: the factory for labels can for example take a vector or a csv file and then returns a labels object, potentially converted. See what I mean?

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: hmm, I see. You mean `sg.machine(ae, sg.layer("NeuralSoftmaxLayer", num_neurons=10), sigma=0.01)` would convert DeepAutoencoder to a NeuralNetwork with an output layer for inference?

**Comment by gf712** [comment]
@karlnapf @vigsterkr shall I merge this and now we have notebooks in CI? It’s an extra 25 minutes of testing but probably still quicker than MSVC build...

File: doc/ipython-notebooks/neuralnets/autoencoders.ipynb (hunk 9 above)
Comment (karlnapf) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: Something in those lines

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: OK. I can look into this in another PR

Comment (karlnapf) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: actually rather than using the factory, we should use the casting functions we have `as_neural_network`, `as_kernel`, etc

Comment (gf712) [reply doc/ipython-notebooks/neuralnets/autoencoders.ipynb:402]: OK, I'll write an issue for this then!

**Comment by gf712** [comment]
Removed PR notebook from buildbot for the time being. Can be brought back when we update the script to only run specific notebooks, rather than using all the donbot cores for each PR :D
